### ISSUE TITLE
Huge memory usage in redis

### ISSUE THREAD

**Issue by mlissner** [freelawproject/courtlistener#1460]
One of the issues we had during the py3 upgrade was that we created about a million failed tasks on the `iauploads` queue. Another issue I've noticed is that redis is using about 15GB of memory, which is way more than it does usually. A third issue I'm aware of from munin is that every so often our disk gets a lot of writes and creates long wait times. 

Putting all of this together, I think that celery is using a huge amount of memory holding failed tasks. It's not configured to do this, but perhaps it's holding them because they're queued for retry and thus haven't fully failed yet.

First step to figuring this out is to look into redis and see what I can learn about what's inside.

**Comment by mlissner** [comment]
Note also that this is a dependency for landing the fix from Flooie having to do with uploading IA data. Those were in #1443 and #1445.

**Comment by mlissner** [comment]
Either figuring out what's inside redis is hard or I'm not very good at it, but this command is a useful one in some ways:

```txt
↪ redis-cli INFO
# Server
redis_version:3.0.5
redis_git_sha1:00000000
redis_git_dirty:0
redis_build_id:b14cbc1543f9ab1a
redis_mode:standalone
os:Linux 4.4.0-193-generic x86_64
arch_bits:64
multiplexing_api:epoll
gcc_version:4.8.4
process_id:27905
run_id:a8bd3b7b6bac7d2caf84eaca0688fd7dca1e8e89
tcp_port:6379
uptime_in_seconds:17039
uptime_in_days:0
hz:10
lru_clock:11325693
config_file:/etc/redis/6379.conf

# Clients
connected_clients:52
client_longest_output_list:0
client_biggest_input_buf:0
blocked_clients:0

# Memory
used_memory:16460919248
used_memory_human:15.33G
used_memory_rss:16670621696
used_memory_peak:16482131496
used_memory_peak_human:15.35G
used_memory_lua:35840
mem_fragmentation_ratio:1.01
mem_allocator:jemalloc-3.6.0

# Persistence
loading:0
rdb_changes_since_last_save:6606
rdb_bgsave_in_progress:1
rdb_last_save_time:1605160887
rdb_last_bgsave_status:ok
rdb_last_bgsave_time_sec:128
rdb_current_bgsave_time_sec:25
aof_enabled:0
aof_rewrite_in_progress:0
aof_rewrite_scheduled:0
aof_last_rewrite_time_sec:-1
aof_current_rewrite_time_sec:-1
aof_last_bgrewrite_status:ok
aof_last_write_status:ok

# Stats
total_connections_received:12821
total_commands_processed:561072
instantaneous_ops_per_sec:66
total_net_input_bytes:2636967159
total_net_output_bytes:418127647
instantaneous_input_kbps:541.80
instantaneous_output_kbps:8.85
rejected_connections:0
sync_full:0
sync_partial_ok:0
sync_partial_err:0
expired_keys:57413
evicted_keys:0
keyspace_hits:20664
keyspace_misses:60328
pubsub_channels:0
pubsub_patterns:2
latest_fork_usec:316591
migrate_cached_sockets:0

# Replication
role:master
connected_slaves:0
master_repl_offset:0
repl_backlog_active:0
repl_backlog_size:1048576
repl_backlog_first_byte_offset:0
repl_backlog_histlen:0

# CPU
used_cpu_sys:69.65
used_cpu_user:84.13
used_cpu_sys_children:858.10
used_cpu_user_children:3744.46

# Cluster
cluster_enabled:0

# Keyspace
db0:keys=196,expires=181,avg_ttl=2116736
db1:keys=241773,expires=241772,avg_ttl=312847684
db2:keys=6579,expires=0,avg_ttl=0
db3:keys=37585,expires=0,avg_ttl=0
db4:keys=5,expires=4,avg_ttl=407952677
```

A few things to note:

1. Note that this doesn't tell you which DB is using the memory, and indeed there's no easy way to do so. 

1. Note in the `keyspace` section that DB 4 is nearly empty and DB 5 is. I thought I had reconfigured these to be the DB's used by the cache and by celery after the upgrade, but it turns out that my approach didn't work at all. 

    My approach was to use 11-private.py to override the REDIS_DATABASES setting. That works in its way, but it happens after the setting is integrated into the celery and cache URLs, so tweaking it in 11-private has no effect. At the moment, these settings cannot be properly overridden because you'd have to do so after they're created and before the URLs are. This could be fixed by creating a 20-public file where the URLs are created, but I'm not sure I care. In the future, the better path is just to override the URLs.

    In any case, these DBs are empty, so we can ignore them.

1. DB 3 is used for sending alerts. It shouldn't have any keys in it normally, but it has a steadily growing number of keys. This needs investigation and I've opened issue #1461 to do so.

1. DB 1 is our cache. It has a lot of keys, but that seems fairly normal. Nevertheless, I used `RANDOMKEY` a bunch of times to note that a big majority of the keys seem to take the form of: `":1:mlt-cluster:4451842"`. That's a cache of MLT results, and it's safe to delete. Unfortunately, there's no easy way to use a wildcard to delete keys, but this worked:

        redis-cli -n 1 --scan --pattern ':1:mlt-cluster*' | xargs redis-cli -n 1 del

    Great. Unfortunately, deleting that huge number of keys didn't fix the memory.

1. I remember from ages ago that memory isn't always reclaimed by redis and that rebooting it was a good way to check that out. Alas, restarting redis didn't help either after the above.

1. DB 2 has stats info. I can't imagine how it'd blow up during a py3 transition. 

So I began to feel like things were getting a bit out of hand. I tried purging the Celery queue with something like:

    cd /var/www/cl/docker/nginx/ && \
        sudo docker-compose exec cl-python celery -Q iauploads purge

That ran for many hours without doing much to my CPU, so I eventually killed it. Something wasn't working there.

A new approach is needed.

**Comment by mlissner** [comment]
I finally realized that there might be a way to find big objects in redis, and discovered this lovely command:

```txt
↪ redis-cli --bigkeys

# Scanning the entire keyspace to find biggest keys as well as
# average sizes per key type.  You can use -i 0.1 to sleep 0.1 sec
# per 100 SCAN commands (not usually needed).

[00.00%] Biggest string found so far 'celery-task-meta-da405b33-fe33-41f2-9204-55654df549f1' with 3026 bytes
[00.00%] Biggest string found so far 'celery-task-meta-3241a393-1385-4285-b83b-519dcab5623c' with 3514 bytes
[16.94%] Biggest set    found so far '_kombu.binding.batch4' with 1 members
[16.94%] Biggest hash   found so far 'unacked' with 2 fields
[16.94%] Biggest set    found so far '_kombu.binding.celeryev' with 1887 members
[50.00%] Biggest zset   found so far 'unacked_index' with 2 members
[58.06%] Biggest string found so far 'celery-task-meta-302499a3-4f2e-4d51-904b-e69ce87b640c' with 3743 bytes
[58.06%] Biggest list   found so far 'celery' with 2 items
[66.13%] Biggest list   found so far 'batch2' with 48406 items
[66.13%] Biggest string found so far 'celery-task-meta-1f948d8b-bc11-4ac8-a946-45f326c82aa2' with 3788 bytes
[83.87%] Biggest list   found so far 'iauploads' with 12244097 items

-------- summary -------

Sampled 124 keys in the keyspace!
Total key length in bytes is 6008 (avg len 48.45)

Biggest string found 'celery-task-meta-1f948d8b-bc11-4ac8-a946-45f326c82aa2' has 3788 bytes
Biggest   list found 'iauploads' has 12244097 items
Biggest    set found '_kombu.binding.celeryev' has 1887 members
Biggest   hash found 'unacked' has 2 fields
Biggest   zset found 'unacked_index' has 2 members

108 strings with 86172 bytes (87.10% of keys, avg size 797.89)
4 lists with 12312054 items (03.23% of keys, avg size 3078013.50)
10 sets with 3762 members (08.06% of keys, avg size 376.20)
1 hashs with 2 fields (00.81% of keys, avg size 2.00)
1 zsets with 2 members (00.81% of keys, avg size 2.00)
```

Note:

    Biggest   list found 'iauploads' has 12244097 items

That looks suspicious. I dug around for a way to zero out a list in Redis, but sort of concluded that ① zeroing out a list in Redis is the same thing as `DEL`'ing it, and ② Celery created the list in the first place and should be able to recreate it if it's missing. 

Finally, I deleted the list with:

    DEL iauploads

And immediately the memory plummeted. 

------

Lessons learned: 

1. `--bigkeys` is the bomb.

1. Can't override redis DB settings.

I think this will do it. Onwards.

**Comment by mlissner** [comment]
I just ran into a variant of this problem. What happened was that while crawling iquery pages, we got blocked from a bunch of courts, that backed up our queue and I had to just nuke the darned thing. To do that, I used `--bigkeys` to find the biggest keys, then I ran `del batch1`. From there, things kept coming in, so I SIGKILL'ed the iquery scraper and iauploader, both of which were running. Seems to have worked. 

Two further thoughts:

1. I am using scheduled tasks a lot for the iquery crawler. I'm not sure if these are still floating around, because they have really long schedules. I'm sort of waiting for them to pop up again, but so far so good.

2. The instructions to use `celery purge` seemed to do nothing. Who knows what's up with that.
